### PR TITLE
Book A Motorcycle - Backend

### PR DESCRIPTION
**Pull request by GhausAliKhan**
### Overview:
This pull request introduces several enhancements and features to the motorcycle booking system, addressing various user and admin functionalities as outlined in the requirements.

### General Requirements:
- Ensure that the `codebase is free of linter errors`. Address and resolve any reported issues to maintain code quality.
- Confirm that the `Gitflow` has been followed correctly. This includes creating feature branches, keeping commits focused and meaningful, and following the defined branching strategy.
- Document work in a professional manner. Include clear and concise explanations of new features, changes, and any considerations for future development in `README.md`
- Implement `Ruby` code following the recommended best practices. This includes adhering to coding conventions, using appropriate naming conventions, and ensuring code readability.

### Key Features: 

- User Authentication:
  Implemented proper user authentication from the `front-end` to the server. Users can now log in by typing their `username`.

- Navigation Panel Updates:
  Added links in the navigation panel for `Motorcycles`, `Reserve`, and `My Reservations` for regular users.
  For `admin` users, included additional links for `Add Motorcycle` and `Delete Motorcycle`.

- Main Page Display:
  Display a list of motorcycles on the main page based on the selected `theme`.

- Motorcycle Details Page:
  Implemented a details page for each motorcycle model, allowing users to view the full description and reserve the motorcycle.

- Reservation Functionality:
  Users can reserve a motorcycle from the details page or access the `Reserve` page from the navigation panel. The `username` and `selected item` are auto-filled for convenience.

  To reserve an appointment, users need to select a `date` and `city`.

- My Reservations:
  Users can view a list of their reservations by clicking the `My Reservations` link in the navigation panel. The list includes information about the `reservation name`, `date`, and `city`.

- Admin Features:
  Admin users can `add` a new motorcycle through the `Add Motorcycle` form.
  Admin users can `view` and `delete` motorcycles through the `Delete Motorcycle` link. Deleting a motorcycle marks it as removed, making it disappear from the main list.

- Responsive Design:
  Created both `mobile` and `desktop` versions to ensure a seamless user experience across different devices.

### Note:
Unfortunately, we faced challenges incorporating API documentation and we were unable to include frontend tests due to a team member's unavailability. Regrettably, time constraints compelled us to prioritize other critical features. We appreciate your understanding and remain committed to addressing this in subsequent updates.

### Testing:
Extensive testing has been conducted to validate the functionality of the added features and ensure a smooth user experience.

### Links: 
- [Kanban Board](https://github.com/GhausAliKhan/Book-A-Motorcycle-Back-End/projects/1)
- [Kanban Board Issue](https://github.com/GhausAliKhan/Book-A-Motorcycle-Back-End/issues/19)
- [Frontend Pull Request](https://github.com/GhausAliKhan/book-a-motorcycle-front-end/pull/7)

### Reviewers:
Your careful review and feedback on the implemented features are highly appreciated.

Thank you for your understanding and collaboration. 💯

### CODE REVIEW

File: spec/requests/api/v1/authentication_spec.rb
Comment (Meltrust) [inline spec/requests/api/v1/authentication_spec.rb:1]: - [ ] Kindly, remove the commented code on this file in order to preserve the repo's cleanliness. 👍

**Review by Meltrust — CHANGES_REQUESTED** [review]
_note: The frontend review will be ready in a few minutes_

Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Tests are passing✔️
-  Nice code organization ✔️ 
-  Good readme ✔️ 




#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

**Review by vichuge — COMMENTED** [review]
Hi dear student, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

star_wars_almost_there

## Suggested changes
_Check the comments under the review._

### Everything is great in backend. please take a look on the [frontend](https://github.com/GhausAliKhan/book-a-motorcycle-front-end/pull/7#pullrequestreview-1857130613)

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by Meltrust — CHANGES_REQUESTED** [review]
Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!






### To highlight:

-  Great job✔️




#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

You can do it

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

**Pull request merged**